File: impl-dragonbox/jcsfloat/jcsfloat.go
// Package jcsfloat implements the ECMAScript Number::toString algorithm for
// IEEE 754 double-precision floating-point values using the Dragonbox algorithm
// (Junekey Jeon, 2020) with fixed-width 128-bit arithmetic.
//
// The algorithm is specified in ECMA-262 §6.1.6.1.20 (Number::toString).
// The digit generation uses precomputed 128-bit powers of 10, carrier-based
// interval decomposition, a shorter-interval optimization for power-of-two
// boundaries, and two-digit-at-a-time extraction for shortest-output with
// ECMA-262 Note 2 (even-digit) tie-breaking.
//
// FormatDouble is validated against large pinned ECMAScript oracle datasets and
// round-trip fuzzing for finite doubles.
package jcsfloat

import (
	"math"
	"math/bits"

	"github.com/lattice-substrate/jcs-dragonbox/jcserr"
)

// FormatDouble formats an IEEE 754 double-precision value exactly as specified
// by the ECMAScript Number::toString algorithm (ECMA-262, radix 10).
//
// ECMA-FMT-001: NaN returns error.
// ECMA-FMT-002: -0 returns "0".
// ECMA-FMT-003: +/-Infinity returns error.
// ECMA-FMT-008: Shortest round-trip representation.
// ECMA-FMT-009: Even-digit tie-breaking.
func FormatDouble(f float64) (string, *jcserr.Error) {
	// ECMA-FMT-001: NaN -> error
	if math.IsNaN(f) {
		return "", jcserr.New(jcserr.InvalidGrammar, -1, "NaN is not representable in JSON")
	}
	// ECMA-FMT-002: -0 and +0 -> "0"
	if f == 0 {
		return "0", nil
	}
	// ECMA-FMT-003: +/-Infinity -> error
	if math.IsInf(f, 0) {
		return "", jcserr.New(jcserr.InvalidGrammar, -1, "Infinity is not representable in JSON")
	}

	negative := false
	if f < 0 {
		negative = true
		f = -f
	}

	digits, dp := dragonboxFloat64(f)
	return formatECMA(negative, digits, dp), nil
}

const (
	mantBits64 = 52
	expBits64  = 11
	bias64     = 1023
	mantMask64 = 1<<mantBits64 - 1
	expMask64  = 1<<expBits64 - 1
)

// dragonboxFloat64 implements Dragonbox for f64 → shortest decimal.
// Returns (digits, dp) where dp is the decimal point position.
//
// The Dragonbox algorithm (Jeon, 2020) differs from Ryu/Schubfach in its
// interval decomposition:
//   - A "shorter interval" optimization for power-of-2 boundaries
//   - Carrier-based arithmetic for the general case
//   - Two-digit-at-a-time extraction in the trimming loop
//
//nolint:gocyclo,cyclop,gocognit,funlen // REQ:ECMA-FMT-008 Dragonbox digit extraction is inherently branch-heavy for shortest-output correctness.
func dragonboxFloat64(f float64) (string, int) {
	fbits := math.Float64bits(f)
	rawMant := fbits & mantMask64
	rawExp := int(fbits>>mantBits64) & expMask64 //nolint:gosec // REQ:ECMA-FMT-008 expMask64 bounds result to [0,2047], safe for int.

	var mant uint64
	var exp int

	if rawExp == 0 {
		// Subnormal
		mant = rawMant
		exp = 1 - bias64 - mantBits64
	} else {
		// Normal
		mant = (1 << mantBits64) | rawMant
		exp = rawExp - bias64 - mantBits64
	}

	// Fast path: exact integer.
	if exp <= 0 && bits.TrailingZeros64(mant) >= -exp {
		m := mant >> uint(-exp) //nolint:gosec // REQ:ECMA-FMT-008 exp<=0 guaranteed by enclosing if; -exp is non-negative.
		return formatUint(m), countDecimalDigits(m)
	}

	// Dragonbox: check for shorter interval case (power-of-2 boundary).
	// When the mantissa is exactly 2^52 (rawMant == 0) and we're not subnormal,
	// the rounding interval is asymmetric and potentially shorter.
	isShorterInterval := rawExp > 1 && rawMant == 0

	if isShorterInterval {
		return dragonboxShorterInterval(mant, exp)
	}

	return dragonboxGeneral(mant, exp)
}

// dragonboxShorterInterval handles the power-of-2 boundary case.
// At these values, the lower half of the rounding interval is narrower,
// and we can sometimes find a shorter representation.
func dragonboxShorterInterval(mant uint64, exp int) (string, int) {
	// For power-of-2 boundaries: interval is (4m-1, 4m, 4m+2) × 2^(e-2).
	ml := 4*mant - 1
	mc := 4 * mant
	mu := 4*mant + 2
	e2 := exp - 2

	if e2 == 0 {
		return dragonboxTrim(ml, mc, mu, true, false, 0)
	}

	q := mulByLog2Log10(-e2) + 1

	var dl, dc, du uint64
	var dl0, dc0, du0 bool

	dl, _, dl0 = mult128bitPow10(ml, e2, q)
	dc, _, dc0 = mult128bitPow10(mc, e2, q)
	du, e2, du0 = mult128bitPow10(mu, e2, q)

	if e2 >= 0 {
		return "0", 0
	}

	if q > 55 {
		dl0, dc0, du0 = false, false, false
	}
	if q < 0 && q >= -24 {
		if divisibleByPower5(ml, -q) {
			dl0 = true
		}
		if divisibleByPower5(mc, -q) {
			dc0 = true
		}
		if divisibleByPower5(mu, -q) {
			du0 = true
		}
	}

	extra := uint(-e2)
	extraMask := uint64(1<<extra - 1)

	fracl := dl & extraMask
	fracc := dc & extraMask
	fracu := du & extraMask
	dl >>= extra
	dc >>= extra
	du >>= extra

	// Accept bounds for even mantissa (Dragonbox: always true at power-of-2 boundary).
	uok := !du0 || fracu > 0
	if du0 && fracu == 0 {
		uok = true // mant&1 == 0 at power-of-2 boundary
	}
	if !uok {
		du--
	}

	var cup bool
	if dc0 {
		cup = fracc > 1<<(extra-1) ||
			(fracc == 1<<(extra-1) && dc&1 == 1)
	} else {
		cup = fracc>>(extra-1) == 1
	}

	lok := dl0 && fracl == 0
	if !lok {
		dl++
	}

	c0 := dc0 && fracc == 0
	return dragonboxTrim(dl, dc, du, c0, cup, q)
}

// dragonboxGeneral handles the common case where the interval is symmetric
// (non-power-of-2 boundary values).
func dragonboxGeneral(mant uint64, exp int) (string, int) {
	// Regular case: (2m-1, 2m, 2m+1) × 2^(e-1).
	ml := 2*mant - 1
	mc := 2 * mant
	mu := 2*mant + 1
	e2 := exp - 1

	if e2 == 0 {
		return dragonboxTrim(ml, mc, mu, true, false, 0)
	}

	q := mulByLog2Log10(-e2) + 1

	var dl, dc, du uint64
	var dl0, dc0, du0 bool

	dl, _, dl0 = mult128bitPow10(ml, e2, q)
	dc, _, dc0 = mult128bitPow10(mc, e2, q)
	du, e2, du0 = mult128bitPow10(mu, e2, q)

	if e2 >= 0 {
		return "0", 0
	}

	if q > 55 {
		dl0, dc0, du0 = false, false, false
	}
	if q < 0 && q >= -24 {
		if divisibleByPower5(ml, -q) {
			dl0 = true
		}
		if divisibleByPower5(mc, -q) {
			dc0 = true
		}
		if divisibleByPower5(mu, -q) {
			du0 = true
		}
	}

	extra := uint(-e2)
	extraMask := uint64(1<<extra - 1)

	fracl := dl & extraMask
	fracc := dc & extraMask
	fracu := du & extraMask
	dl >>= extra
	dc >>= extra
	du >>= extra

	acceptBounds := mant&1 == 0
	uok := !du0 || fracu > 0
	if du0 && fracu == 0 {
		uok = acceptBounds
	}
	if !uok {
		du--
	}

	var cup bool
	if dc0 {
		cup = fracc > 1<<(extra-1) ||
			(fracc == 1<<(extra-1) && dc&1 == 1)
	} else {
		cup = fracc>>(extra-1) == 1
	}

	lok := dl0 && fracl == 0 && acceptBounds
	if !lok {
		dl++
	}

	c0 := dc0 && fracc == 0
	return dragonboxTrim(dl, dc, du, c0, cup, q)
}

// dragonboxTrim performs Dragonbox's digit trimming with two-digit-at-a-time
// extraction where possible.
// Returns (digits, dp) where dp is the decimal point position.
//
//nolint:gocyclo,cyclop // REQ:ECMA-FMT-008 Dragonbox two-digit trimming loop.
func dragonboxTrim(dl, dc, du uint64, c0, cup bool, q int) (string, int) {
	trimmed := 0
	cNextDigit := uint64(0)

	// Dragonbox optimization: try removing two digits at a time.
	for du >= 100 {
		l := (dl + 99) / 100
		c := dc / 100
		u := du / 100
		if l > u {
			break
		}
		// Recover the two removed digits for rounding.
		removed := dc - c*100
		if l == c+1 && c < u {
			c++
			removed = 0
			cup = false
		}
		trimmed += 2
		c0 = c0 && cNextDigit == 0 && removed/10 == 0
		cNextDigit = removed / 10
		dl, dc, du = l, c, u
	}

	// Single-digit trimming for the remainder.
	for du > 0 {
		l := (dl + 9) / 10
		c := dc / 10
		cdigit := dc % 10
		u := du / 10
		if l > u {
			break
		}
		if l == c+1 && c < u {
			c++
			cdigit = 0
			cup = false
		}
		trimmed++
		c0 = c0 && cNextDigit == 0
		cNextDigit = cdigit
		dl, dc, du = l, c, u
	}

	// Round-to-even tie-breaking (ECMA-FMT-009).
	if trimmed > 0 {
		cup = cNextDigit > 5 ||
			(cNextDigit == 5 && !c0) ||
			(cNextDigit == 5 && c0 && dc&1 == 1)
	}
	if dc < du && cup {
		dc++
	}

	// Render digits.
	var buf [20]byte
	n := 20
	v := dc
	for v > 0 {
		n--
		buf[n] = byte(v%10) + '0'
		v /= 10
	}
	if n == 20 {
		n = 19
		buf[19] = '0'
	}

	ndigits := 20 - n

	// Trim trailing zeros.
	end := 20
	for end > n+1 && buf[end-1] == '0' {
		end--
		trimmed++
	}

	dp := ndigits + trimmed - q
	return string(buf[n:end]), dp
}

// mulByLog2Log10 returns floor(x * log(2)/log(10)).
func mulByLog2Log10(x int) int {
	return (x * 78913) >> 18
}

// mulByLog10Log2 returns floor(x * log(10)/log(2)).
func mulByLog10Log2(x int) int {
	return (x * 108853) >> 15
}

// mult128bitPow10 multiplies m by 10^q using 128-bit table lookup.
func mult128bitPow10(m uint64, e2, q int) (uint64, int, bool) {
	if q == 0 {
		return m << 8, e2 - 8, true
	}
	if q < pow10MinExp || pow10MaxExp < q {
		return 0, 0, false
	}
	pow := pow10Table[q-pow10MinExp]
	if q < 0 {
		pow[0]++
	}
	e2 += mulByLog10Log2(q) - 127 + 119

	l1, l0 := bits.Mul64(m, pow[0])
	h1, h0 := bits.Mul64(m, pow[1])
	mid, carry := bits.Add64(l1, h0, 0)
	h1 += carry
	return h1<<9 | mid>>55, e2, mid<<9 == 0 && l0 == 0
}

// divisibleByPower5 reports whether m is divisible by 5^k.
func divisibleByPower5(m uint64, k int) bool {
	if m == 0 {
		return true
	}
	for i := 0; i < k; i++ {
		if m%5 != 0 {
			return false
		}
		m /= 5
	}
	return true
}

// formatUint formats a uint64 as a decimal string (trailing zeros trimmed).
func formatUint(v uint64) string {
	if v == 0 {
		return "0"
	}
	var buf [20]byte
	n := 20
	for v > 0 {
		n--
		buf[n] = byte(v%10) + '0'
		v /= 10
	}
	end := 20
	for end > n+1 && buf[end-1] == '0' {
		end--
	}
	return string(buf[n:end])
}

// countDecimalDigits returns the number of decimal digits in v.
func countDecimalDigits(v uint64) int {
	if v == 0 {
		return 1
	}
	n := 0
	for v > 0 {
		n++
		v /= 10
	}
	return n
}

File: impl-dragonbox-rs/src/jcsfloat.rs
use crate::jcserr::{FailureClass, JcsError};
use crate::jcsfloat_format::format_ecma;
use crate::pow10_table::{POW10_MAX_EXP, POW10_MIN_EXP, POW10_TABLE};

const MANT_BITS_64: u32 = 52;
const EXP_BITS_64: u32 = 11;
const BIAS_64: i32 = 1023;
const MANT_MASK_64: u64 = (1u64 << MANT_BITS_64) - 1;
const EXP_MASK_64: u32 = (1u32 << EXP_BITS_64) - 1;

/// Formats an IEEE 754 double-precision value exactly as specified by the
/// ECMAScript Number::toString algorithm (ECMA-262, radix 10).
///
/// Uses the Dragonbox algorithm (Junekey Jeon, 2020) with carrier-based
/// interval decomposition and two-digit-at-a-time extraction.
pub fn format_double(f: f64) -> Result<String, JcsError> {
    if f.is_nan() {
        return Err(JcsError::new(
            FailureClass::InvalidGrammar,
            -1,
            "NaN is not representable in JSON",
        ));
    }
    if f == 0.0 {
        return Ok("0".to_string());
    }
    if f.is_infinite() {
        return Err(JcsError::new(
            FailureClass::InvalidGrammar,
            -1,
            "Infinity is not representable in JSON",
        ));
    }

    let negative;
    let fabs;
    if f < 0.0 {
        negative = true;
        fabs = -f;
    } else {
        negative = false;
        fabs = f;
    }

    let (digits, dp) = dragonbox_float64(fabs);
    Ok(format_ecma(negative, &digits, dp))
}

fn dragonbox_float64(f: f64) -> (String, i32) {
    let fbits = f.to_bits();
    let raw_mant = fbits & MANT_MASK_64;
    let raw_exp = ((fbits >> MANT_BITS_64) as u32) & EXP_MASK_64;

    let mant: u64;
    let exp: i32;

    if raw_exp == 0 {
        mant = raw_mant;
        exp = 1 - BIAS_64 - MANT_BITS_64 as i32;
    } else {
        mant = (1u64 << MANT_BITS_64) | raw_mant;
        exp = raw_exp as i32 - BIAS_64 - MANT_BITS_64 as i32;
    }

    if exp <= 0 && mant.trailing_zeros() >= (-exp) as u32 {
        let m = mant >> ((-exp) as u32);
        return (format_uint(m), count_decimal_digits(m));
    }

    // Dragonbox: shorter interval case at power-of-2 boundaries.
    let is_shorter_interval = raw_exp > 1 && raw_mant == 0;
    if is_shorter_interval {
        dragonbox_shorter_interval(mant, exp)
    } else {
        dragonbox_general(mant, exp)
    }
}

fn dragonbox_shorter_interval(mant: u64, exp: i32) -> (String, i32) {
    let ml = 4 * mant - 1;
    let mc = 4 * mant;
    let mu = 4 * mant + 2;
    let e2 = exp - 2;

    if e2 == 0 {
        return dragonbox_trim(ml, mc, mu, true, false, 0);
    }

    let q = mul_by_log2_log10(-e2) + 1;

    let (dl, _, dl0_init) = mult_128bit_pow10(ml, e2, q);
    let (dc, _, dc0_init) = mult_128bit_pow10(mc, e2, q);
    let (du, e2_final, du0_init) = mult_128bit_pow10(mu, e2, q);

    if e2_final >= 0 {
        return ("0".to_string(), 0);
    }

    let mut dl0 = dl0_init;
    let mut dc0 = dc0_init;
    let mut du0 = du0_init;

    if q > 55 {
        dl0 = false;
        dc0 = false;
        du0 = false;
    }
    if (-24..0).contains(&q) {
        if divisible_by_power5(ml, (-q) as u32) {
            dl0 = true;
        }
        if divisible_by_power5(mc, (-q) as u32) {
            dc0 = true;
        }
        if divisible_by_power5(mu, (-q) as u32) {
            du0 = true;
        }
    }

    let extra = (-e2_final) as u32;
    let extra_mask = (1u64 << extra) - 1;

    let fracl = dl & extra_mask;
    let fracc = dc & extra_mask;
    let fracu = du & extra_mask;
    let mut dl = dl >> extra;
    let dc = dc >> extra;
    let mut du = du >> extra;

    let mut uok = !du0 || fracu > 0;
    if du0 && fracu == 0 {
        uok = true; // mant&1 == 0 at power-of-2 boundary
    }
    if !uok {
        du -= 1;
    }

    let cup = if dc0 {
        fracc > (1u64 << (extra - 1))
            || (fracc == (1u64 << (extra - 1)) && dc & 1 == 1)
    } else {
        (fracc >> (extra - 1)) == 1
    };

    let lok = dl0 && fracl == 0;
    if !lok {
        dl += 1;
    }

    let c0 = dc0 && fracc == 0;
    dragonbox_trim(dl, dc, du, c0, cup, q)
}

fn dragonbox_general(mant: u64, exp: i32) -> (String, i32) {
    let ml = 2 * mant - 1;
    let mc = 2 * mant;
    let mu = 2 * mant + 1;
    let e2 = exp - 1;

    if e2 == 0 {
        return dragonbox_trim(ml, mc, mu, true, false, 0);
    }

    let q = mul_by_log2_log10(-e2) + 1;

    let (dl, _, dl0_init) = mult_128bit_pow10(ml, e2, q);
    let (dc, _, dc0_init) = mult_128bit_pow10(mc, e2, q);
    let (du, e2_final, du0_init) = mult_128bit_pow10(mu, e2, q);

    if e2_final >= 0 {
        return ("0".to_string(), 0);
    }

    let mut dl0 = dl0_init;
    let mut dc0 = dc0_init;
    let mut du0 = du0_init;

    if q > 55 {
        dl0 = false;
        dc0 = false;
        du0 = false;
    }
    if (-24..0).contains(&q) {
        if divisible_by_power5(ml, (-q) as u32) {
            dl0 = true;
        }
        if divisible_by_power5(mc, (-q) as u32) {
            dc0 = true;
        }
        if divisible_by_power5(mu, (-q) as u32) {
            du0 = true;
        }
    }

    let extra = (-e2_final) as u32;
    let extra_mask = (1u64 << extra) - 1;

    let fracl = dl & extra_mask;
    let fracc = dc & extra_mask;
    let fracu = du & extra_mask;
    let mut dl = dl >> extra;
    let dc = dc >> extra;
    let mut du = du >> extra;

    let accept_bounds = mant & 1 == 0;
    let mut uok = !du0 || fracu > 0;
    if du0 && fracu == 0 {
        uok = accept_bounds;
    }
    if !uok {
        du -= 1;
    }

    let cup = if dc0 {
        fracc > (1u64 << (extra - 1))
            || (fracc == (1u64 << (extra - 1)) && dc & 1 == 1)
    } else {
        (fracc >> (extra - 1)) == 1
    };

    let lok = dl0 && fracl == 0 && accept_bounds;
    if !lok {
        dl += 1;
    }

    let c0 = dc0 && fracc == 0;
    dragonbox_trim(dl, dc, du, c0, cup, q)
}

/// Dragonbox digit trimming with two-digit-at-a-time extraction.
fn dragonbox_trim(
    mut dl: u64,
    mut dc: u64,
    mut du: u64,
    mut c0: bool,
    mut cup: bool,
    q: i32,
) -> (String, i32) {
    let mut trimmed: i32 = 0;
    let mut c_next_digit: u64 = 0;

    // Two-digit-at-a-time optimization.
    while du >= 100 {
        let l = dl.div_ceil(100);
        let c = dc / 100;
        let u = du / 100;
        if l > u {
            break;
        }
        let removed = dc - c * 100;
        if l == c + 1 && c < u {
            dc = c + 1;
            dl = l;
            du = u;
            c_next_digit = 0;
            cup = false;
            trimmed += 2;
            c0 = c0 && c_next_digit == 0 && removed / 10 == 0;
            continue;
        }
        trimmed += 2;
        c0 = c0 && c_next_digit == 0 && removed / 10 == 0;
        c_next_digit = removed / 10;
        dl = l;
        dc = c;
        du = u;
    }

    // Single-digit trimming.
    while du > 0 {
        let l = dl.div_ceil(10);
        let c = dc / 10;
        let cdigit = dc % 10;
        let u = du / 10;
        if l > u {
            break;
        }
        if l == c + 1 && c < u {
            dc = c + 1;
            dl = l;
            du = u;
            c_next_digit = 0;
            cup = false;
            trimmed += 1;
            c0 = c0 && c_next_digit == 0;
            continue;
        }
        trimmed += 1;
        c0 = c0 && c_next_digit == 0;
        c_next_digit = cdigit;
        dl = l;
        dc = c;
        du = u;
    }

    if trimmed > 0 {
        cup = c_next_digit > 5
            || c_next_digit == 5 && !c0
            || c_next_digit == 5 && dc & 1 == 1;
    }
    if dc < du && cup {
        dc += 1;
    }

    let mut buf = [0u8; 20];
    let mut n = 20usize;
    let mut v = dc;
    while v > 0 {
        n -= 1;
        buf[n] = b'0' + (v % 10) as u8;
        v /= 10;
    }
    if n == 20 {
        n = 19;
        buf[19] = b'0';
    }

    let ndigits = (20 - n) as i32;

    let mut end = 20usize;
    while end > n + 1 && buf[end - 1] == b'0' {
        end -= 1;
        trimmed += 1;
    }

    let dp = ndigits + trimmed - q;
    let digits = std::str::from_utf8(&buf[n..end]).unwrap().to_string();
    (digits, dp)
}

fn mul_by_log2_log10(x: i32) -> i32 {
    ((x as i64 * 78913) >> 18) as i32
}

fn mul_by_log10_log2(x: i32) -> i32 {
    ((x as i64 * 108853) >> 15) as i32
}

fn mult_128bit_pow10(m: u64, e2: i32, q: i32) -> (u64, i32, bool) {
    if q == 0 {
        return (m << 8, e2 - 8, true);
    }
    if !(POW10_MIN_EXP..=POW10_MAX_EXP).contains(&q) {
        return (0, 0, false);
    }
    let idx = (q - POW10_MIN_EXP) as usize;
    let mut pow = POW10_TABLE[idx];
    if q < 0 {
        pow.0 = pow.0.wrapping_add(1);
    }
    let e2_out = e2 + mul_by_log10_log2(q) - 127 + 119;

    let full_lo = m as u128 * pow.0 as u128;
    let full_hi = m as u128 * pow.1 as u128;
    let l1 = (full_lo >> 64) as u64;
    let l0 = full_lo as u64;
    let h1 = (full_hi >> 64) as u64;
    let h0 = full_hi as u64;
    let (mid, carry) = l1.overflowing_add(h0);
    let h1 = h1 + carry as u64;

    (
        (h1 << 9) | (mid >> 55),
        e2_out,
        (mid << 9) == 0 && l0 == 0,
    )
}

fn divisible_by_power5(mut m: u64, k: u32) -> bool {
    if m == 0 {
        return true;
    }
    for _ in 0..k {
        if m % 5 != 0 {
            return false;
        }
        m /= 5;
    }
    true
}

fn format_uint(mut v: u64) -> String {
    if v == 0 {
        return "0".to_string();
    }
    let mut buf = [0u8; 20];
    let mut n = 20usize;
    while v > 0 {
        n -= 1;
        buf[n] = b'0' + (v % 10) as u8;
        v /= 10;
    }
    let mut end = 20usize;
    while end > n + 1 && buf[end - 1] == b'0' {
        end -= 1;
    }
    std::str::from_utf8(&buf[n..end]).unwrap().to_string()
}

fn count_decimal_digits(mut v: u64) -> i32 {
    if v == 0 {
        return 1;
    }
    let mut n = 0i32;
    while v > 0 {
        n += 1;
        v /= 10;
    }
    n
}
